### DISCUSSION TITLE
Does libvips work asynchronously with php?

### DISCUSSION THREAD

**Discussion by IOrlandoni** [libvips/php-vips#158] [Q&A]
I'm doing a small image conversion.  png to jpg

$x = vips_image_new_from_file("X.PNG",['access' => 'sequential'])["out"];
 
 try {
if(vips_image_write_to_file($x,"indir/$filename.JPG",["Q"=>$kalite])==True) {
 echo " Converted to successfully  "; 
 } else {  throw new Exception("File writing problem i am not sync");  }   
 } 
 catch(Exception $e) {
  echo 'Message: ' .$e->getMessage();
}

I'm doing a small image conversion. However, I think libvips works asynchronously and does not block php. So throw new Exception("File writing problem i am not sync"); I get an error.

How can I wait for this process to finish?
PHP 8.1 JIT enabled

**Comment by jcupitt** [comment]
Hello @web-apply,

It looks like you are using the old php-vips-ext module -- this is not supported! Please use [php-vips](https://github.com/libvips/php-vips).

php-vips is synchronous, so (for example) `$image->writeToFile("x.jpg");` will block until the file is written.

You'd need several processes if you want asynchronous operations, with perhaps a job queue. I've used DB tables to implement this in the last, perhaps there's something better now.

**Comment by rlerdorf** [comment]
Have a look at Gearman (PHP ext is https://github.com/php/pecl-networking-gearman)
You normally don't want your frontend web servers grinding away at cpu-heavy jobs. You are better off asynchronously sending those tasks off to a pool of servers. That is exactly what Gearman is designed to do. You can still use php-vips to actually do the work on those gearman servers, of course. If you are tiny and just have a single server, you can still run gearman locally to manage these async jobs. That allows for very easy scaling as you grow.
